firedrake/functionspacedata.py

```diff
@@ -369,6 +373,64 @@ def entity_dofs_key(entity_dofs):
     return key
 
 
+def entity_permutations_key(entity_permutations):
+    """Provide a canonical key for an entity_permutations dict.
+
+    :arg entity_permutations: The FInAT entity_permutations.
+    :returns: A tuple of canonicalised entity_permutations (suitable for
+        caching).
+    """
+    key = []
+    for k in sorted(entity_permutations.keys()):
+        sub_key = [k]
+        for sk in sorted(entity_permutations[k]):
+            subsub_key = [sk]
+            for ssk in sorted(entity_permutations[k][sk]):
+                subsub_key.append((ssk, tuple(entity_permutations[k][sk][ssk])))
+            sub_key.append(tuple(subsub_key))
+        key.append(tuple(sub_key))
+    key = tuple(key)
+    return key
+
+
+import ufl
+from tsfc.finatinterface import create_element
+def get_shared_data_key(mesh, element, use_true_tensor_element_shape=False):
```

Comment:
We almost can use the same code in the constructor of `FunctionSpaceData` to get `(nodes_per_entity, real_tensorproduct)` cache key, but this function wants to take `ufl_element` as argument, while the constructor of `FunctionSpaceData` takes `finat_element`. I think it simplifies the code around `FunctionSpace` construction if we make the constructor of `FunctionSpaceData` (and thus `get_shared_data`) take `ufl_element` instead of `finat_element`. If that makes sense, I can make a small PR which just does it.

Reply:
OK, I think that seems reasonable.

Reply:
#2154 has been merged and I've cleaned up these lines.